****Penerapan Algoritma _Clustering_ untuk Pengelompokkan Saham IDX Berdasarkan Indikator-indikator Fundamental - Submission Machine Learning Terapan Dicoding**** 

oleh: Fikri Septrian Anggara (fikri_anggara_2c3r)

### Import library yang diperlukan

In [2]:
# untuk pengolahan data
import pandas as pd

# untuk visualisasi data
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme()

# untuk pembangunan klaster
from sklearn.cluster import KMeans
from sklearn.cluster import AgglomerativeClustering

# untuk visualisasi klaster
from scipy.cluster.hierarchy import dendrogram, linkage

# untuk scaling data
from sklearn.preprocessing import MinMaxScaler

# untuk menghitung jarak
from scipy.spatial.distance import pdist, cdist

# untuk menghitung evaluasi klaster
from sklearn.metrics import silhouette_score

# untuk melakukan reduksi dimensi
from sklearn.decomposition import PCA

# 2. Data Understanding

### 2.1. Menyiapkan Dataset
dataset diperoleh dari kaggle, terdapat dua dataset yang digunakan yaitu:
 - [financial statement idx stocks](https://www.kaggle.com/datasets/greegtitan/financial-statement-idx-stocks?resource=download) (kaggle). terakhir diupdate pada Oktober 2022.
 - [daftar saham](https://www.kaggle.com/datasets/muamkh/ihsgstockdata?select=DaftarSaham.csv) (kaggle). terakhir diupdate pada Januari 2023

In [3]:
# load data
stockQuarter = pd.read_csv('data/quarter.csv')
masterStock = pd.read_csv('data/DaftarSaham.csv')

### 2.2. Overview Data
pada tahap ini dilakukan _overview_ pada data stockQuarter dan masterStock

In [4]:
# banyaknya baris dan kolom (baris, kolom)
print(stockQuarter.shape)
print(masterStock.shape)

(208691, 8)
(829, 14)


In [5]:
# kolom dan tipe data
stockQuarter.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208691 entries, 0 to 208690
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   symbol      208691 non-null  object 
 1   account     208691 non-null  object 
 2   type        208149 non-null  object 
 3   2021-09-30  51722 non-null   float64
 4   2021-12-31  51538 non-null   float64
 5   2022-03-31  53449 non-null   float64
 6   2022-06-30  50375 non-null   float64
 7   2022-09-30  17013 non-null   float64
dtypes: float64(5), object(3)
memory usage: 12.7+ MB


In [6]:
# kolom dan tipe data
masterStock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 829 entries, 0 to 828
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Code                829 non-null    object 
 1   Name                829 non-null    object 
 2   ListingDate         829 non-null    object 
 3   Shares              829 non-null    float64
 4   ListingBoard        829 non-null    object 
 5   Sector              829 non-null    object 
 6   LastPrice           824 non-null    float64
 7   MarketCap           824 non-null    float64
 8   MinutesFirstAdded   787 non-null    object 
 9   MinutesLastUpdated  787 non-null    object 
 10  HourlyFirstAdded    806 non-null    object 
 11  HourlyLastUpdated   806 non-null    object 
 12  DailyFirstAdded     824 non-null    object 
 13  DailyLastUpdated    824 non-null    object 
dtypes: float64(3), object(11)
memory usage: 90.8+ KB


In [7]:
# ambil sampel random sebanyak 5 data
stockQuarter.sample(n=5, random_state=1)


symbol account                                         type  \
181194   SQMI      CF  Cash From Discontinued Investing Activities   
137781   MSKY      CF                Paymentson Behalfof Employees   
22706    BBHI      BS  Long Term Debt And Capital Lease Obligation   
143195   NETV      IS                                  Credit Card   
184286   SUPR      BS                   Securities And Investments   

          2021-09-30    2021-12-31    2022-03-31    2022-06-30    2022-09-30  
181194           NaN           NaN           NaN           NaN           NaN  
137781 -8.421300e+10 -7.747800e+10 -6.249900e+10 -6.643200e+10           NaN  
22706   4.114605e+09  3.291684e+09  0.000000e+00  4.575037e+10  4.135613e+10  
143195           NaN           NaN           NaN           NaN           NaN  
184286           NaN           NaN           NaN           NaN           NaN

In [8]:
stockQuarter.isnull().sum()

symbol             0
account            0
type             542
2021-09-30    156969
2021-12-31    157153
2022-03-31    155242
2022-06-30    158316
2022-09-30    191678
dtype: int64

- terdapat banyak ****null value**** pada data saham perkuarter

In [10]:
# ambil sampel random sebanyak 5 data
masterStock.sample(n=5, random_state=1)

Code                        Name          ListingDate        Shares  \
370  INTA          Intraco Penta Tbk.           1993-08-23  3.343935e+09   
449  LION       Lion Metal Works Tbk.           1993-08-20  5.201600e+08   
810  WINR  Winner Nusantara Jaya Tbk.           2022-04-25  5.235200e+09   
578  PDPP  Primadaya Plastisindo Tbk.  2022-11-09 00:00:00  2.500000e+09   
700  SMRU              SMR Utama Tbk.           2011-10-10  1.249939e+10   

     ListingBoard                    Sector  LastPrice     MarketCap  \
370  Pengembangan               Industrials       74.0  2.474512e+11   
449         Utama               Industrials      735.0  3.823176e+11   
810         Utama  Properties & Real Estate       50.0  2.617600e+11   
578  Pengembangan           Basic Materials      202.0  5.050000e+11   
700  Pengembangan                    Energy       50.0  6.249693e+11   

       MinutesFirstAdded   MinutesLastUpdated     HourlyFirstAdded  \
370  2021-11-01 09:00:00  2022-06-03 15:59:00  2020-04-16 09:00:00   
449  2021-11-01 09:00:00  2023-01-06 15:59:00  2020-04-16 09:00:00   
810  2022-04-25 09:00:00  2023-01-06 15:59:00  2022-04-25 09:00:00   
578  2022-11-09 09:00:00  2023-01-06 15:59:00  2022-11-09 09:00:00   
700                  NaN                  NaN                  NaN   

       HourlyLastUpdated DailyFirstAdded DailyLastUpdated  
370  2022-06-03 16:00:00      2001-04-16       2023-01-06  
449  2023-01-06 15:00:00      2001-04-16       2023-01-06  
810  2023-01-06 15:00:00      2022-04-25       2023-01-06  
578  2023-01-06 15:00:00      2022-11-10       2023-01-06  
700                  NaN      2011-10-10       2023-01-06

In [11]:
# 5 number summary stock quarter
stockQuarter.describe(include='all').T

count unique                top   freq                  mean  \
symbol       208691    544               AALI    391                   NaN   
account      208691      3                 BS  73461                   NaN   
type         208149    388  Interest Received   1080                   NaN   
2021-09-30  51722.0    NaN                NaN    NaN  1918202605901.322021   
2021-12-31  51538.0    NaN                NaN    NaN  2042450921973.788574   
2022-03-31  53449.0    NaN                NaN    NaN  1940242135220.246094   
2022-06-30  50375.0    NaN                NaN    NaN  2139261720062.068848   
2022-09-30  17013.0    NaN                NaN    NaN  3359261415683.166504   

                              std                min      25%            50%  \
symbol                        NaN                NaN      NaN            NaN   
account                       NaN                NaN      NaN            NaN   
type                          NaN                NaN      NaN            NaN   
2021-09-30  24798786684018.976562 -175313000000000.0      0.0   5837091260.0   
2021-12-31  26106014954333.765625 -438813037000000.0      0.0   7402772575.0   
2022-03-31  25585023607145.355469 -223695701000000.0      0.0   5526650874.0   
2022-06-30  26835599887835.515625 -194169000000000.0      0.0   7726820540.0   
2022-09-30  33720778198892.175781 -197134246000000.0  79886.0  20220521000.0   

                       75%                 max  
symbol                 NaN                 NaN  
account                NaN                 NaN  
type                   NaN                 NaN  
2021-09-30  233651664187.5  1637950171000000.0  
2021-12-31  257243562462.0  1725611128000000.0  
2022-03-31  214030861495.0  1734074740000000.0  
2022-06-30  265889690000.0  1785706841000000.0  
2022-09-30  557420000000.0  1839336498000000.0

- terdapat **544** buah saham
- terdapat **388** variabel pada laporan keuangan
- terdapat **3** kategori akun yaitu balance sheet, cash-flow dan income statement

In [12]:
# 5 number summary master stock
masterStock.describe(include='all').T

count unique                      top freq  \
Code                  829    829                     AALI    1   
Name                  829    829  Astra Agro Lestari Tbk.    1   
ListingDate           829    713      2022-11-08 00:00:00    6   
Shares              829.0    NaN                      NaN  NaN   
ListingBoard          829      3             Pengembangan  422   
Sector                829     11       Consumer Cyclicals  142   
LastPrice           824.0    NaN                      NaN  NaN   
MarketCap           824.0    NaN                      NaN  NaN   
MinutesFirstAdded     787     60      2021-11-01 09:00:00  698   
MinutesLastUpdated    787     18      2023-01-06 15:59:00  759   
HourlyFirstAdded      806    106      2020-04-16 09:00:00  662   
HourlyLastUpdated     806     15      2023-01-06 15:00:00  759   
DailyFirstAdded       824    501               2001-04-16  225   
DailyLastUpdated      824      2               2023-01-06  822   

                                     mean                    std  \
Code                                  NaN                    NaN   
Name                                  NaN                    NaN   
ListingDate                           NaN                    NaN   
Shares                 10946944360.291918     45395931792.267258   
ListingBoard                          NaN                    NaN   
Sector                                NaN                    NaN   
LastPrice                     1363.453883            3733.432446   
MarketCap           10076631834955.722656  50531421356601.734375   
MinutesFirstAdded                     NaN                    NaN   
MinutesLastUpdated                    NaN                    NaN   
HourlyFirstAdded                      NaN                    NaN   
HourlyLastUpdated                     NaN                    NaN   
DailyFirstAdded                       NaN                    NaN   
DailyLastUpdated                      NaN                    NaN   

                             min             25%              50%  \
Code                         NaN             NaN              NaN   
Name                         NaN             NaN              NaN   
ListingDate                  NaN             NaN              NaN   
Shares                 3600000.0    1230839821.0     3128090000.0   
ListingBoard                 NaN             NaN              NaN   
Sector                       NaN             NaN              NaN   
LastPrice                   25.0          101.75            287.0   
MarketCap           9000000000.0  297528301025.0  1111304868160.0   
MinutesFirstAdded            NaN             NaN              NaN   
MinutesLastUpdated           NaN             NaN              NaN   
HourlyFirstAdded             NaN             NaN              NaN   
HourlyLastUpdated            NaN             NaN              NaN   
DailyFirstAdded              NaN             NaN              NaN   
DailyLastUpdated             NaN             NaN              NaN   

                                75%                 max  
Code                            NaN                 NaN  
Name                            NaN                 NaN  
ListingDate                     NaN                 NaN  
Shares                 9327465018.0     1184363929502.0  
ListingBoard                    NaN                 NaN  
Sector                          NaN                 NaN  
LastPrice                     975.0             38000.0  
MarketCap           4236952973400.0  1012951085850000.0  
MinutesFirstAdded               NaN                 NaN  
MinutesLastUpdated              NaN                 NaN  
HourlyFirstAdded                NaN                 NaN  
HourlyLastUpdated               NaN                 NaN  
DailyFirstAdded                 NaN                 NaN  
DailyLastUpdated                NaN                 NaN

- Terdapat 11 sektor pada master stok dengan **Consumer Cyclicals** adalah sektor yang paling banyak emitennya
- Terdapat 829 buah emiten

In [13]:
# variabel pada laporan keuangan
pd.DataFrame(pd.unique(stockQuarter['type']))

0
0                          Long Term Equity Investment
1                           Additional Paid In Capital
2                                       Long Term Debt
3                               Ordinary Shares Number
4                                         Total Assets
..                                                 ...
384                      Interest Income From Deposits
385                            Credit Losses Provision
386                              Other Interest Income
387  Interest Expense For Federal Funds Sold And Se...
388                     Total Money Market Investments

[389 rows x 1 columns]

_dataframe_ **stockQuarter** memiliki 208691 baris dan 8 kolom. ke delapan kolom yaitu :
 - **symbol**: Kode saham IDX seperti BBRI, BBCA, BMRI, dst.
 - **account**: Akun laporan keuangan. nilainya meliputi **BS** untuk _Balance Sheet_, **IS** untuk akun _Income Statement_, dan **CF** untuk akun _Cash Flow_.
 - **type**: Tipe/variabel laporan keuangan seperti data total aset, kredit, dividen yang dibayarkan, dst. memiliki 388 tipe.
 - Kolom data variabel laporan keuangan perkuarter. meliputi tanggal **2021-09-30**, **2021-12-31**, **2022-03-31**, **2022-06-30**, **2022-09-30**.
 
_dataframe_ **masterStock** memiliki 829 baris dan 14 kolom. ke empat belas kolom tersebut yaitu:
 - **Code**: Kode saham IDX
 - **Name**: Nama saham
 - **ListingDate**: Tanggal pendaftaran saham
 - **Shares**: Total saham beredar
 - **ListingBoard**: Tingkat pasar saham, meliputi tingkat akselerasi, pengembangan dan utama
 - **Sector**: Sektor perusahaan
 - **LastPrice**: Harga terakhir saham
 - **MarketCap**: Total nilai perusahaan
 - **MinutesFirstAdded**: Menit pertama data ditambahkan
 - **MinutesLastUpdated**: Menit terakhir data diperbarui
 - **HourlyFirstAdded**: Jam pertama data ditambahkan
 - **HourlyLastUpdated**: Jam terakhir data diperbarui
 - **DailyFirstAdded**: Tanggal Pertama data ditambahkan
 - **DailyLastUpdated**: Tanggal Terakhir data diperbarui

terdapat 389 variabel pada laporan keuangan

Berdasarkan _Overview Data_, diketahui :
- Terdapat banyak **null value** pada data saham perkuarter
- Terdapat **544** buah saham yang tercatat laporan keuangannya
- Terdapat **388** variabel pada laporan keuangan
- Terdapat **3** kategori akun yaitu balance sheet, cash-flow dan income statement
- Terdapat **11** sektor pada master stok dengan **Consumer Cyclicals** adalah sektor yang paling banyak emitennya
- Terdapat total **829** buah saham
- Terdapat **389** variabel pada laporan keuangan


# 3. Data Preparation

Data stockQuarter masih belum memiliki struktur yang bisa digunakan untuk pembuatan model dan belum digabung dengan masterStock untuk memperoleh data sektor. 
Maka pada ada tahap _Data Preparation_, penulis :
1. merubah struktur data stockQuarter agar cocok untuk pembangunan model klaster
2. menggabungkan data stockQuarter dan masterStock
3. melakukan feature engineering untuk memperoleh indikator fundamental yang digunakan pada paper Examining the effectiveness of fundamental analysis
in a long-term stock portfolio
4. menghilangkan fitur yang tidak digunakan
5. melakukan analisis data eksploratif
6. melakukan imputasi pada fitur hasil feature engineering
7. melakukan reduksi dimensi 

karena data stockQuarter terbaru (2022-09-30) memiliki paling banyak _null value_, maka penulis menggunakan data kuarter sebelumnya, yaitu data kuarter kedua tahun 2022.

In [14]:
# buat kolom baru, gabungan antara account dan type untuk menyimpan nilai tipe akun
stockQuarter['account_type'] = stockQuarter['account']+'_'+stockQuarter['type']
stockQuarter.head()

symbol account                         type    2021-09-30    2021-12-31  \
0   AALI      BS  Long Term Equity Investment  3.235200e+11  3.309040e+11   
1   AALI      BS   Additional Paid In Capital  3.878995e+12  3.878995e+12   
2   AALI      BS               Long Term Debt  5.709887e+12  2.131944e+12   
3   AALI      BS       Ordinary Shares Number  1.924688e+09  1.924688e+09   
4   AALI      BS                 Total Assets  2.969401e+13  3.039991e+13   

     2022-03-31    2022-06-30    2022-09-30                    account_type  
0  3.275800e+11  2.850690e+11  4.714630e+11  BS_Long Term Equity Investment  
1  3.878995e+12  3.878995e+12  3.878995e+12   BS_Additional Paid In Capital  
2  2.144732e+12  2.220370e+12  3.281008e+12               BS_Long Term Debt  
3  1.924688e+09  1.924688e+09  1.924688e+09       BS_Ordinary Shares Number  
4  3.123278e+13  3.023399e+13  3.263865e+13                 BS_Total Assets

### 3.1 Mengubah Struktur Data

In [15]:
# ambil data 2022 kuarter dua saja
data2022 = stockQuarter[['symbol', 'account_type', '2022-06-30']].copy()

# reshaping data, kolom" data yang baru merupakan value dari 'type' dari df yang lama
dataReshaped = pd.DataFrame(data2022.pivot_table(
    index='symbol',
    columns='account_type',
    values='2022-06-30'
).reset_index())
dataReshaped.sample(n=5, random_state=1)

account_type symbol  BS_Accounts Payable  BS_Accounts Receivable  \
306            META         2.169320e+10            2.629671e+10   
342            NFCX         2.339294e+11            1.993821e+11   
47             BACA         4.088400e+10                     NaN   
67             BEBS         4.705185e+09            8.377203e+10   
376            PRIM         2.716511e+10            6.700829e+10   

account_type  BS_Accumulated Depreciation  BS_Additional Paid In Capital  \
306                         -1.550597e+11                   4.691558e+11   
342                         -5.261832e+10                   2.834290e+11   
47                          -3.065800e+11                   8.488000e+09   
67                          -1.148769e+10                   9.190475e+10   
376                         -1.453821e+11                   4.610570e+11   

account_type  BS_Allowance For Doubtful Accounts Receivable  \
306                                             -87700000.0   
342                                            -320225284.0   
47                                                      NaN   
67                                             -357552323.0   
376                                                     NaN   

account_type  BS_Allowance For Loans And Lease Losses  \
306                                               NaN   
342                                               NaN   
47                                       1.073200e+10   
67                                                NaN   
376                                               NaN   

account_type  BS_Assets Held For Sale  BS_Available For Sale Securities  \
306                               NaN                               NaN   
342                               NaN                      3.359086e+10   
47                                NaN                      2.928573e+12   
67                                NaN                               NaN   
376                               NaN                               NaN   

account_type  BS_Buildings And Improvements  ...  \
306                            1.301870e+11  ...   
342                            1.418633e+11  ...   
47                             4.072670e+11  ...   
67                             3.858019e+10  ...   
376                            4.198495e+11  ...   

account_type  IS_Selling General And Administration  \
306                                    9.808912e+09   
342                                    6.519324e+09   
47                                     2.162000e+10   
67                                     2.794760e+09   
376                                    5.785167e+09   

account_type  IS_Special Income Charges  IS_Tax Effect Of Unusual Items  \
306                                 NaN                             0.0   
342                                 NaN                             0.0   
47                                  NaN                             0.0   
67                                  NaN                             0.0   
376                                 NaN                             0.0   

account_type  IS_Tax Provision  IS_Tax Rate For Calcs  \
306               1.088595e+10                    0.0   
342               1.711056e+09                    0.0   
47                9.570000e+08                    0.0   
67                1.036365e+10                    0.0   
376               1.650462e+09                    0.0   

account_type  IS_Total Premiums Earned  IS_Total Revenue  \
306                                NaN      2.043580e+11   
342                                NaN      2.428054e+12   
47                                 NaN      1.195280e+11   
67                                 NaN      1.607964e+11   
376                                NaN      6.797255e+10   

account_type  IS_Total Unusual Items  \
306                              NaN   
342                              NaN   
47                               NaN   
67

In [16]:
# saham yang tersedia
pd.DataFrame(pd.unique(dataReshaped['symbol']))

0
0    AALI
1    ABBA
2    ABDA
3    ABMM
4    ACES
..    ...
498  WSKT
499  WTON
500  YPAS
501  ZBRA
502  ZINC

[503 rows x 1 columns]

In [17]:
# cek null value dari masing masing kolom
nulltable = pd.DataFrame(dataReshaped.isnull().sum().reset_index().iloc[1:])
nulltable.columns = ['financial statement', 'sum of null']
print(nulltable)
print('total null: ', dataReshaped.isnull().sum().sum())

                               financial statement  sum of null
1                              BS_Accounts Payable           17
2                           BS_Accounts Receivable           70
3                      BS_Accumulated Depreciation           10
4                    BS_Additional Paid In Capital            5
5    BS_Allowance For Doubtful Accounts Receivable          162
..                                             ...          ...
222                       IS_Total Premiums Earned          482
223                               IS_Total Revenue           14
224                         IS_Total Unusual Items          182
225      IS_Total Unusual Items Excluding Goodwill          182
226                                   IS_Write Off          398

[226 rows x 2 columns]
total null:  63737


In [18]:
# 5 data teratas
dataReshaped.head()

account_type symbol  BS_Accounts Payable  BS_Accounts Receivable  \
0              AALI         8.259340e+11            4.305280e+11   
1              ABBA         4.565925e+10            1.927453e+10   
2              ABDA         1.726138e+10            7.472096e+10   
3              ABMM         1.617792e+08            1.994225e+08   
4              ACES         1.728831e+11            7.668345e+10   

account_type  BS_Accumulated Depreciation  BS_Additional Paid In Capital  \
0                           -1.276215e+13                   3.878995e+12   
1                           -1.616807e+11                  -4.555371e+10   
2                           -1.051394e+11                   8.109426e+09   
3                           -7.378516e+08                   1.150872e+08   
4                           -1.918547e+12                   4.405749e+11   

account_type  BS_Allowance For Doubtful Accounts Receivable  \
0                                             -2.553900e+10   
1                                             -6.364641e+10   
2                                                       NaN   
3                                             -1.031222e+07   
4                                              5.312285e+06   

account_type  BS_Allowance For Loans And Lease Losses  \
0                                                 NaN   
1                                                 NaN   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   

account_type  BS_Assets Held For Sale  BS_Available For Sale Securities  \
0                                 NaN                               NaN   
1                                 NaN                      1.659867e+11   
2                                 NaN                     -2.732850e+09   
3                                 NaN                      2.346308e+07   
4                                 NaN                               NaN   

account_type  BS_Buildings And Improvements  ...  \
0                              4.876408e+12  ...   
1                              2.361747e+10  ...   
2                              9.103365e+10  ...   
3                              8.629613e+07  ...   
4                              2.333466e+12  ...   

account_type  IS_Selling General And Administration  \
0                                      1.818320e+11   
1                                      4.808846e+09   
2                                      2.573074e+10   
3                                      9.630067e+06   
4                                      8.751484e+10   

account_type  IS_Special Income Charges  IS_Tax Effect Of Unusual Items  \
0                          2.423000e+09                    7.975096e+08   
1                          0.000000e+00                    0.000000e+00   
2                          5.790000e+06                    2.316000e+06   
3                                   NaN                    0.000000e+00   
4                         -6.539966e+09                   -1.060284e+09   

account_type  IS_Tax Provision  IS_Tax Rate For Calcs  \
0                 1.628460e+11                    0.0   
1                 3.356401e+08                    0.0   
2                -7.352070e+08                    0.0   
3                 2.569592e+07                    0.0   
4                 1.740837e+10                    0.0   

account_type  IS_Total Premiums Earned  IS_Total Revenue  \
0                                  NaN      4.383457e+12   
1                                  NaN      3.270153e+10   
2                         1.570100e+11      1.933188e+11   
3                                  NaN      3.655084e+08   
4                                  NaN      1.681350e+12   

account_type  IS_Total Unusual Items  \
0                       2.423000e+09   
1                       0.000000e+00   
2                       5.790000e+06   
3 

### 3.2. Menggabungkan Data 

In [21]:
# filter kolom master data
filteredMasterStock = masterStock[['Code', 'Name', 'Shares', 'Sector', 'LastPrice']]

# ubah nama kolom agar bisa dijoin
renamedData = dataReshaped.rename(columns={'symbol':'Code'})

# join data stockQuarter dan masterStock
joinedData = pd.merge(left=renamedData, right=filteredMasterStock, on='Code', how='inner')

In [28]:
joinedData

Code  BS_Accounts Payable  BS_Accounts Receivable  \
0    AALI         8.259340e+11            4.305280e+11   
1    ABBA         4.565925e+10            1.927453e+10   
2    ABDA         1.726138e+10            7.472096e+10   
3    ABMM         1.617792e+08            1.994225e+08   
4    ACES         1.728831e+11            7.668345e+10   
..    ...                  ...                     ...   
498  WSKT         7.007696e+12            1.501147e+13   
499  WTON         2.239787e+12            1.119072e+12   
500  YPAS         1.161044e+10            8.208517e+10   
501  ZBRA         5.400567e+11            6.963487e+11   
502  ZINC         1.101933e+11            5.717424e+10   

     BS_Accumulated Depreciation  BS_Additional Paid In Capital  \
0                  -1.276215e+13                   3.878995e+12   
1                  -1.616807e+11                  -4.555371e+10   
2                  -1.051394e+11                   8.109426e+09   
3                  -7.378516e+08                   1.150872e+08   
4                  -1.918547e+12                   4.405749e+11   
..                           ...                            ...   
498                -4.096478e+12                   1.379202e+13   
499                -1.791698e+12                   9.886334e+11   
500                -1.646578e+11                   2.805402e+10   
501                -3.866211e+11                   1.167524e+12   
502                -4.867618e+11                   3.220000e+10   

     BS_Allowance For Doubtful Accounts Receivable  \
0                                    -2.553900e+10   
1                                    -6.364641e+10   
2                                              NaN   
3                                    -1.031222e+07   
4                                     5.312285e+06   
..                                             ...   
498                                  -2.903400e+12   
499                                  -9.098793e+10   
500                                  -8.574100e+08   
501                                  -3.877312e+09   
502                                            NaN   

     BS_Allowance For Loans And Lease Losses  BS_Assets Held For Sale  \
0                                        NaN                      NaN   
1                                        NaN                      NaN   
2                                        NaN                      NaN   
3                                        NaN                      NaN   
4                                        NaN                      NaN   
..                                       ...                      ...   
498                                      NaN                      NaN   
499                                      NaN                      NaN   
500                                      NaN                      NaN   
501                                      NaN                      NaN   
502                                      NaN                      NaN   

     BS_Available For Sale Securities  BS_Buildings And Improvements  ...  \
0                                 NaN                   4.876408e+12  ...   
1                        1.659867e+11                   2.361747e+10  ...   
2                       -2.732850e+09                   9.103365e+10  ...   
3                        2.346308e+07                   8.629613e+07  ...   
4                                 NaN                   2.333466e+12  ...   
..                                ...                            ...  ...   
498                      8.829152e+11                   1.591137e+12  ...   
499                               NaN                   3.351810e+11  ...   
500                               NaN                   6.509308e+10  ...   
501                               NaN                   1.430180e+11  ...   
502                               NaN                   7.055236e+10  ...   

     IS_Total Revenue  IS_Total Unusual Items  \
0  

### 3.3. _Feature Engineering_ indikator fundamental
berdasarkan paper [[5]](fdsafdsavfdsa), terdapat 5 indikator keuangan yang mampu mewakili indikator lain, yaitu:
- **Net Profit Margin**: perbandingan antara net profit/income dengan total revenue. Melihat apakah pengelolaan perusahaan menghasilkan cukup laba dan apakah biaya operasional dan apakah terdapat biaya yang berlebihan.
- **Debt to equity ratio (D/E)** : perbandingan antara total kewajiban (_liabilities_) dengan ekuitas pemegang saham (_shareholder equity_)
- **Current Ratio**: perbandingan antara aset yang dimiliki dengan kewajiban. menunjukkan kemampuan perusahaan melunasi utang jangka pendek dengan aset lancarnya.
- **Earning per share (EPS)**: perbandingan laba/profit (net income) setelah dikurangi pajak dengan jumlah saham yang beredar (outstanding shares). Digunakan untuk melihat profitabilitas perusahaan. Outstanding Shares diperoleh dari Share Issued - Treasury Shares Number.
- **P/E Ratio**: perbandingan antara harga perlembar saham dengan laba tahunan perlembar (EPS). Untuk membandingkan nilai relatif antar perusahaan.

In [29]:
# Net profit margin = net income / total revenue
joinedData['Net_Profit_Margin'] = joinedData['IS_Net Income']/joinedData['IS_Total Revenue']

# debt to equity ratio (D/E)= total debt / stockholder equity
joinedData['Debt_to_Equity_Ratio'] = joinedData['BS_Total Debt']+joinedData['BS_Stockholders Equity']

# Current Ratio = current Assets (cash dan cash equivalents, accounts receivables, Available For Sale Securities)
# / Current liability ( di kasus ini hanya account payable, Current Notes Payable, Income Tax Payable, Trading Liabilities)
joinedData['Curent_Ratio'] = (joinedData['BS_Cash And Cash Equivalents']+joinedData['BS_Accounts Receivable']+joinedData['BS_Available For Sale Securities'])/(joinedData['BS_Accounts Payable']+joinedData['BS_Current Notes Payable']+joinedData['BS_Income Tax Payable']+joinedData['BS_Trading Liabilities'])

# EPS = Net Income / (share issued - treasury shares number)
joinedData['Earning_per_Shares'] = joinedData['IS_Net Income']/(joinedData['BS_Share Issued']-joinedData['BS_Treasury Shares Number'])

# P/E ratio = share price/EPS
joinedData['Price_to_EPS'] = joinedData['LastPrice']/joinedData['Earning_per_Shares']


KeyError: 'BS_Trading Liabilities'

In [20]:
# usedColumns = [k for k in dataReshaped.columns.tolist() if 'BS_' in k][:15]


In [22]:
# dataFinal = dataReshaped[usedColumns]
# dataFinal

account_type  BS_Accounts Payable  BS_Accounts Receivable  \
0                    8.259340e+11            4.305280e+11   
1                    4.565925e+10            1.927453e+10   
2                    1.726138e+10            7.472096e+10   
3                    1.617792e+08            1.994225e+08   
4                    1.728831e+11            7.668345e+10   
..                            ...                     ...   
498                  7.007696e+12            1.501147e+13   
499                  2.239787e+12            1.119072e+12   
500                  1.161044e+10            8.208517e+10   
501                  5.400567e+11            6.963487e+11   
502                  1.101933e+11            5.717424e+10   

account_type  BS_Accumulated Depreciation  BS_Additional Paid In Capital  \
0                           -1.276215e+13                   3.878995e+12   
1                           -1.616807e+11                  -4.555371e+10   
2                           -1.051394e+11                   8.109426e+09   
3                           -7.378516e+08                   1.150872e+08   
4                           -1.918547e+12                   4.405749e+11   
..                                    ...                            ...   
498                         -4.096478e+12                   1.379202e+13   
499                         -1.791698e+12                   9.886334e+11   
500                         -1.646578e+11                   2.805402e+10   
501                         -3.866211e+11                   1.167524e+12   
502                         -4.867618e+11                   3.220000e+10   

account_type  BS_Allowance For Doubtful Accounts Receivable  \
0                                             -2.553900e+10   
1                                             -6.364641e+10   
2                                                       NaN   
3                                             -1.031222e+07   
4                                              5.312285e+06   
..                                                      ...   
498                                           -2.903400e+12   
499                                           -9.098793e+10   
500                                           -8.574100e+08   
501                                           -3.877312e+09   
502                                                     NaN   

account_type  BS_Allowance For Loans And Lease Losses  \
0                                                 NaN   
1                                                 NaN   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   
..                                                ...   
498                                               NaN   
499                                               NaN   
500                                               NaN   
501                                               NaN   
502                                               NaN   

account_type  BS_Assets Held For Sale  BS_Available For Sale Securities  \
0                                 NaN                               NaN   
1                                 NaN                      1.659867e+11   
2                                 NaN                     -2.732850e+09   
3                                 NaN                      2.346308e+07   
4                                 NaN                               NaN   
..                                ...                               ...   
498                               NaN                      8.829152e+11   
499                               NaN                               NaN   
500                               NaN                               NaN   
501                               NaN                               NaN   
502                               NaN                               NaN   

account_typ